File: src/forbidden_stdlib_ffi.erl
-module(forbidden_stdlib_ffi).
-export([is_truthy/1, coerce/1, yeet/1, procrastinate/0, nan/0, is_nan/1, rescue/1, defer/2, on_crash/2]).

is_truthy(X) ->
    X /= false.

coerce(X) ->
    X.

yeet(X) ->
    throw(X).

procrastinate() ->
    _Hmm = 1 + 1,
    procrastinate().

nan() ->
    nan.

is_nan(N) ->
    nan == N.

% stolen from https://github.com/lpil/exception/blob/37d8ef257e9bb0b9e8c6963747a498ca5b6e2099/src/exception_ffi.erl
rescue(F) ->
    try {ok, F()}
    catch
      error:Term -> {error, {errored, Term}};
      throw:Term -> {error, {thrown, Term}};
      exit:Term -> {error, {exited, Term}}
    end.

defer(Cleanup, Body) ->
    try Body()
    after Cleanup()
    end.

on_crash(Cleanup, Body) ->
    try Body()
    catch A:B:C ->
        Cleanup(), 
        erlang:raise(A,B,C)
    end.

File: src/forbidden_stdlib_ffi.mjs
import { Ok, Error as GError } from "./gleam.mjs";
import { Errored, Thrown } from "./forbidden/exception.mjs";

export function is_truthy(x) {
    return !!x
}

export function coerce(x) {
    return x
}
export function yeet(x) {
    throw x
}
export function procrastinate() {
    while (true) {
        const hmm = 41 + 1
        hmm
    }
}

export function nan() {
    return NaN
}
export function is_nan(n) {
    return isNaN(n)
}

// stolen from https://github.com/lpil/exception/blob/37d8ef257e9bb0b9e8c6963747a498ca5b6e2099/src/exception_ffi.mjs

export function on_crash(cleanup, body) {
    try {
        return body();
    } catch (e) {
        cleanup();
        throw e;
    }
}

export function rescue(f) {
    try {
        return new Ok(f());
    } catch (e) {
        if (e instanceof Error) {
            return new GError(new Errored(e));
        } else {
            return new GError(new Thrown(e));
        }
    }
}

export function defer(cleanup, body) {
    try {
        return body();
    } finally {
        cleanup();
    }
}

export function labor(length) {
    const start = Date.now()
    while (start + length > Date.now()) {
        () => { }
    }
}

export function evaluate(code) {
    return eval(code)
}
